### ISSUE TITLE
Poor quality of default KrigingAlgorithm on Branin test-case

### ISSUE THREAD

**Issue by regislebrun** [openturns/openturns#1738]
See in [this preprint](https://deepai.org/publication/numerical-issues-in-maximum-likelihood-parameter-estimation-for-gaussian-process-regression) an interesting comparison of different Python kriging packages on the Branin test case. It is OT 1.13 which is tested, and the results are not so good for OT... But the tested version is quite old, maybe things are not so bad now.

**Comment by KieranDelamotte** [comment]
Is this example useful ?

http://openturns.github.io/openturns/master/auto_meta_modeling/kriging_metamodel/plot_kriging_branin_function.html?highlight=branin

**Comment by regislebrun** [comment]
For sure ;-)!

**Comment by KieranDelamotte** [comment]
That being said, I think there is a small issue there as the variance map I get from the previous example is this:

![Figure_3](https://user-images.githubusercontent.com/10544677/106033051-8cb47480-60d1-11eb-9e63-9485decfb770.png)

We see that variance at the training data is *really* small. When I print the variance at the 5 first training point I get:

[2.84217e-14,1.42109e-14,2.84217e-14,4.26326e-14,6.1594e-27]

and not what is displayed in the current example.

**Comment by regislebrun** [comment]
@KieranDelamotte There is nothing wrong with the variance. The conditional variance is expected to be zero at the learning points, and it is essentially what you get here. Taking the square root, you get standard deviations of the order of 1e-7, and as the standard deviation landscape is nearly flat on most of the domain (meaning that the kriging approximation is quite good), there is no isoline on most of the graph with the levels you chose. I tweaked your script a little bit and got the following graph:
![std_2d](https://user-images.githubusercontent.com/8220933/106119033-05a8e000-6155-11eb-9f5d-4805cf431486.png)

In order to better see if there are steep valleys I exported the standard deviation graph as a VTK file and Paraview gives this vizualisation:
![std](https://user-images.githubusercontent.com/8220933/106118933-eca02f00-6154-11eb-815a-93d03ccd63a8.png)

It confirms that the standard deviation is small everywhere but in the corners, forming steep valleys. You can see (with good eyes) that the lowest value is of order 1e-5 on the grid vertices.

**Comment by KieranDelamotte** [comment]
Hi,

Thanks for your input  @regislebrun . I know the variance at the training points should be null and I agree that 1e-7 might be considered as null in this context.
What I do not understand is that the *same* example script executed with my compiled version gives 

[2.84217e-14,1.42109e-14,2.84217e-14,4.26326e-14,6.1594e-27]

but on the published documentation it is 

[0.0158562,0.0134627,0.0081106,0.00950279,0.0110715].

**Comment by jschueller** [comment]
maybe some resourcemap key is altered in some other graph ? you could try ResourceMap.Reload but it's broken on master and needs my pr

**Comment by mbaudin47** [comment]
Thank you for sharing the paper. The conclusions may be worth reading:

![image](https://user-images.githubusercontent.com/31351465/106157799-5f280380-6183-11eb-895d-728e8d75bb96.png)

**Comment by mbaudin47** [comment]
I tried to understand how we may use `MultiStart` on the Branin test case with current OT. This was important to better see how #1788 using `MultiStart` with QMC design points internally would compare to #1797 improvement of the `MultiStart` constructor. I also wanted to see what exactly is wrong with Branin test case.

The following script uses the usecases module to get the branin_function. Then I fit a kriging using two different methods:
* using the default settings (i.e. uses a local solver),
* using a `MultiStart` on QMC batteries.

The issue is straightforward: one single optimization generally fail. Hence, I catch the exceptions and repeat the attempt until it succeeds. 

```
import openturns as ot
from openturns.usecases import branin_function as branin_function
ot.RandomGenerator_SetSeed(1976)
ot.Log.Show(ot.Log.NONE)
# Load the Branin model
bm = branin_function.BraninModel()

model = bm.objectiveFunction

marginal_collection = [ot.Uniform(bm.lowerbound[i], bm.upperbound[i]) for i in range(bm.dim)]
distribution = ot.ComposedDistribution(marginal_collection)

number_of_tries = 20
sampleSize_train = 10

for i in range(number_of_tries):
    print("Attempt #", i)
    try:
        X_train = distribution.getSample(sampleSize_train)
        Y_train = model(X_train)
        
        basis = ot.ConstantBasisFactory(bm.dim).build()
        covarianceModel = ot.SquaredExponential([1.0]*bm.dim, [1.0])
        covarianceModel.setScale(X_train.getMax())  # Trick A
        algo = ot.KrigingAlgorithm(X_train, Y_train, covarianceModel, basis)
        scaleOptimizationBounds = ot.Interval(bm.lowerbound, bm.upperbound)
        algo.setOptimizationBounds(scaleOptimizationBounds)  # Trick B
        algo.run()
        result = algo.getResult()
        krigingMetamodel = result.getMetaModel()
        
        basic_covariance_model = result.getCovarianceModel()
        print("Basic covariance model:")
        print(basic_covariance_model)
        print("    Success: quit.")
        break
    except TypeError as err:
        print("    Exception:", err)

# Sobol' sequence
sequence = ot.SobolSequence(bm.dim)
sample_size = 10
experiment = ot.LowDiscrepancyExperiment(sequence, distribution, sample_size, True)
startingPoints = experiment.generate()

# Multistart
local_solver = ot.TNC()
multiStartSolver = ot.MultiStart(local_solver, startingPoints)

# Set multistart into kriging
algo = ot.KrigingAlgorithm(X_train, Y_train, covarianceModel, basis)
algo.setOptimizationAlgorithm(multiStartSolver)
algo.setOptimizationBounds(scaleOptimizationBounds)  # Trick B
algo.run()

finetune_covariance_model = result.getCovarianceModel()
print("Fine tuned covariance model:")
print(finetune_covariance_model)
```

Here is the output:
```
Attempt # 0
    Exception: InvalidArgumentException : In CovarianceModelImplementation::setParameter, the component 0 of scale is non positive
Attempt # 1
    Exception: InvalidArgumentException : In CovarianceModelImplementation::setParameter, the component 1 of scale is non positive
[...]
Attempt # 6
Basic covariance model:
SquaredExponential(scale=[0.36226,0.82518], amplitude=[2.60813])
    Success: quit.
Fine tuned covariance model:
SquaredExponential(scale=[0.36226,0.82518], amplitude=[2.60813])
```

We see that the 6th DOE with a local solver produced the same covariance model as one single DOE with a  `MultiStart`. More precisely, the failures are generated because of zeros scale values of the scale parameter during MLE.

@josephmure: I guess that PR #1797 would be used in this particular test case this way:
```
number_of_local_starting_points = 20  # The value suggested in (Basak, Petit, Bect, Vazquez, 2021)
local_solver = ot.TNC()
global_solver = ot.MultiStart(local_solver, number_of_local_starting_points)
algokriging.setOptimizationAlgorithm(global_solver)
```

Am I correct?

**Comment by regislebrun** [comment]
You should use another local solver than TNC. From my experience this solver is fragile with kriging (maybe due to poor gradients?) and quite costly. Cobyla is a better choice, in particular because you don't expect to have an 8 digits optimal point.

**Comment by JPelamatti** [comment]
I am not sure how to address this issue.

There are a few examples of how to parameterize a multi-start optimization of the likelihood :
https://openturns.github.io/openturns/latest/auto_meta_modeling/kriging_metamodel/plot_kriging_hyperparameters_optimization.html?highlight=multistart
https://openturns.github.io/openturns/latest/auto_meta_modeling/kriging_metamodel/plot_kriging_categorical.html?highlight=multistart
https://openturns.github.io/openturns/latest/auto_meta_modeling/kriging_metamodel/plot_kriging_advanced.html?highlight=multistart

I believe the only real issue remains the definition of the distribution of the starting hyperparameters sample, for which I still disagree with the method proposed in our examples

**Comment by JPelamatti** [comment]
I am taking the liberty of closing this issue as the problem is not very clear. 
We can however agree that the use of multi-start optimization of the likelihood should be a generalized practice when using OT.
